ngraph_bridge/ngraph_data_cache.h

```diff
@@ -0,0 +1,178 @@
+  std::deque<KeyType> m_lru;
+  int m_depth;
+  absl::Mutex m_mutex;
+
+  // Test class
+  friend class tensorflow::ngraph_bridge::testing::
+      NGraphDataCacheTest_SameKeyMultiThread_Test;
+  friend class tensorflow::ngraph_bridge::testing::
+      NGraphDataCacheTest_RemoveItemTest_Test;
+};
+
+template <typename KeyType, typename ValueType>
+NgraphDataCache<KeyType, ValueType>::NgraphDataCache(int depth)
+    : m_depth(depth) {}
+
+template <typename KeyType, typename ValueType>
+NgraphDataCache<KeyType, ValueType>::~NgraphDataCache() {
+  m_ng_items_map.clear();
+}
+
+template <typename KeyType, typename ValueType>
+void NgraphDataCache<KeyType, ValueType>::RemoveItem(
+    KeyType key, std::function<void(ValueType)> callback_destroy_item) {
+  if (m_ng_items_map.find(key) != m_ng_items_map.end()) {
+    callback_destroy_item(m_ng_items_map.at(key));
+    RemoveItem(key);
+  }
+}
+
+template <typename KeyType, typename ValueType>
+void NgraphDataCache<KeyType, ValueType>::RemoveItem(KeyType key) {
+  if (m_ng_items_map.find(key) != m_ng_items_map.end()) {
+    absl::MutexLock lock(&m_mutex);
+    m_ng_items_map.erase(key);
+    m_lru.pop_back();
+  }
+}
+
+template <typename KeyType, typename ValueType>
+void NgraphDataCache<KeyType, ValueType>::RemoveAll(
+    std::function<void(ValueType)> callback_destroy_item) {
+  absl::MutexLock lock(&m_mutex);
+  for (auto it = m_ng_items_map.begin(); it != m_ng_items_map.end(); it++) {
+    callback_destroy_item(it->second);
+  }
+  m_ng_items_map.erase(m_ng_items_map.begin(), m_ng_items_map.end());
+}
+
+template <typename KeyType, typename ValueType>
+std::pair<Status, ValueType>
+NgraphDataCache<KeyType, ValueType>::LookUpOrCreate(
+    KeyType key,
+    std::function<std::pair<Status, ValueType>()> callback_create_item,
+    std::function<void(ValueType)> callback_destroy_item) {
+  bool found_in_cache;
+  // look up in the cache
+  {
+    absl::MutexLock lock(&m_mutex);
+    auto it = m_ng_items_map.find(key);
+    found_in_cache = (it != m_ng_items_map.end());
+    if (found_in_cache) {
+      return std::make_pair(Status::OK(), m_ng_items_map.at(key));
+    }
+  }
+  // Item not found in cache, create item
+  ValueType item;
+  auto status_item_pair = callback_create_item();
+
+  // If item is successfully created we will place in the cache.
+  if (status_item_pair.first == Status::OK()) {
+    item = status_item_pair.second;
+    ValueType item_to_evict;
+    bool need_to_evict = false;
+    // lock begins
+    {
+      absl::MutexLock lock(&m_mutex);
+      // Remove item if cache is full
+      if (m_ng_items_map.size() == m_depth) {
+        auto key_to_evict = m_lru.back();
+        item_to_evict = m_ng_items_map.at(key_to_evict);
+        m_ng_items_map.erase(key_to_evict);
+        m_lru.pop_back();
+        need_to_evict = true;
+      }
+      // Add item to cache
+      auto it = m_ng_items_map.emplace(key, item);
+      if (it.second == true) {
+        m_lru.push_front(key);
```

Comment:
if item was added we move the key to front.
if item was not added, ie. the key was already there, still the key needs to be moved to front, since this has been recently used?

Reply:
The lru needs to be updated. dont see that happening

Reply:
Agree, Done.